### PR TITLE
Updated ad_group_check to search groups properly

### PR DESCRIPTION
**Pull request by justinIs**
Added 2 config attributes:
- user_lookup_attribute
- group_lookup_attribute

### CODE REVIEW

**Comment by mikekovacevic** [comment]
Unit tests?

File: lib/devise_ldap_authenticatable/ldap/connection.rb [hunk 1]
```diff
@@ -20,15 +20,20 @@ def initialize(params = {})
...
+        @user_lookup_attribute = ldap_config["user_lookup_attribute"] || 'mail'
+        @group_lookup_attribute = ldap_config["group_lookup_attribtue"] || 'memberof'
         @group_base = ldap_config["group_base"]
         @check_group_membership = ldap_config.has_key?("check_group_membership") ? ldap_config["check_group_membership"] : ::Devise.ldap_check_group_membership
+        @check_group_membership_without_admin = ldap_config.has_key?("check_group_membership_without_admin") ? ldap_config["check_group_membership_without_admin"] : ::Devise.ldap_check_group_membership_without_adminkj
         @required_groups = ldap_config["required_groups"]
         @required_attributes = ldap_config["require_attribute"]
 
+        @user_lookup_attribute = ldap_config["user_lookup_attribute"] || 'mail'
+        @group_lookup_attribute = ldap_config["group_lookup_attribtue"] || 'memberof'
```
Comment (justinIs) [inline lib/devise_ldap_authenticatable/ldap/connection.rb]: @user_lookup_attribute and @group_lookup_attribute are declared above as well. please remove one of them

File: lib/devise_ldap_authenticatable/ldap/connection.rb [hunk 2]
```diff
@@ -122,31 +127,44 @@ def in_required_groups?
...
+          # # AD optimization - extension will recursively check sub-groups with one query
+          # # "(memberof:1.2.840.113556.1.4.1941:=group_name)"
+          # search_result = group_checking_ldap.search(:base => dn,
+          #                   :filter => Net::LDAP::Filter.ex("memberof:1.2.840.113556.1.4.1941", group_name),
+          #                   :scope => Net::LDAP::SearchScope_BaseObject)
+          # # Will return  the user entry if belongs to group otherwise nothing
+          # #if search_result.length == 1 && search_result[0].dn.eql?(dn)
+          # #  in_group = true
+          #   DeviseLdapAuthenticatable::Logger.send("User #{dn} IS included in group: #{group_name}")
+          # end
```
Comment (justinIs) [inline lib/devise_ldap_authenticatable/ldap/connection.rb]: let's remove the commented out code, we can always refer to git history or the original version of this code if we need to

File: lib/devise_ldap_authenticatable/ldap/connection.rb [hunk 3]
```diff
@@ -122,31 +127,44 @@ def in_required_groups?
...
+          filter = Net::LDAP::Filter.join(
+          Net::LDAP::Filter.eq(@user_lookup_attribute, dn),
+          Net::LDAP::Filter.eq(@group_lookup_attribute, group_name)
+          )
+          #filter = Net::LDAP::Filter.eq(@user_lookup_attribute, dn)
```
Comment (justinIs) [inline lib/devise_ldap_authenticatable/ldap/connection.rb]: same here

File: lib/devise_ldap_authenticatable/ldap/connection.rb [hunk 4]
```diff
@@ -122,31 +127,44 @@ def in_required_groups?
...
+          search_result = group_checking_ldap.search(base: @ldap.base, filter: filter, return_result: true, attributes: %w[memberOf])
+          puts @user_lookup_attribute, dn, search_result.inspect
```
Comment (justinIs) [inline lib/devise_ldap_authenticatable/ldap/connection.rb]: remove this `puts` line

File: spec/unit/connection_spec.rb [hunk 5]
```diff
@@ -2,13 +2,40 @@
...
+            'attribute' => 'cn',
+            'user_lookup_attribute' => 'mail',
+            'group_lookup_attribute' => 'memberof',
+            'admin_user' => 'cn=admin,dc=test,dc=com',
+            'admin_password' => 'secret'
+        }}
+        ::Devise.ldap_ad_group_check = true
+        ::Devise.ldap_auth_username_builder = Proc.new() {|attribute, login, ldap| "#{attribute}=#{login},#{ldap.base}"}
+        connection = Devise::LDAP::Connection.new(:login => admin.email, :password => admin.password, :admin => true)
+        connection.in_group? ''
```
Comment (justinIs) [inline spec/unit/connection_spec.rb]: we need to do a little more here, our tests should specify a group_name to do a lookup on.

We also need to test the different cases, like checking a user that IS in a particular group, and checking a user that ISN'T in a particular group, etc.

**Comment by justinIs** [comment]
👍

**Pull request merged**